From 90efc7a06e93b36535700a3034a7b992aee65dff Mon Sep 17 00:00:00 2001
From: ooxi <violetland@mail.ru>
Date: Tue, 7 Jan 2020 18:40:32 +0100
Subject: [PATCH] Print serial number of CM0+ target

By printing the serial number of the target acted upon enables a lot of use cases like statistical analysis and tracability. The serial number is read from the devices memory using SWD.

Successfully tested on samd21 hardware as well as verified using datasheet of the following targets:

* samd09
* samd10
* samd11
* samd20
* samd21
* samc21
* saml21
* saml22
* samr21
* samr30
* samda1

However the targets samr34 as well as samr35 do not seem to support serial number extraction.
---
 target_atmel_cm0p.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/target_atmel_cm0p.c b/target_atmel_cm0p.c
index dc26b77..795fb28 100644
--- a/target_atmel_cm0p.c
+++ b/target_atmel_cm0p.c
@@ -66,6 +66,11 @@
 #define DSU_STATUSA_CRSTEXT    (1 << 1)
 #define DSU_STATUSB_PROT       (1 << 0)
 
+#define SERIAL_NUMBER_0        0x0080a00c
+#define SERIAL_NUMBER_1        0x0080a040
+#define SERIAL_NUMBER_2        0x0080a044
+#define SERIAL_NUMBER_3        0x0080a048
+
 #define NVMCTRL_CTRLA          0x41004000
 #define NVMCTRL_CTRLB          0x41004004
 #define NVMCTRL_PARAM          0x41004008
@@ -175,6 +180,7 @@ static void finish_reset(void)
 static void target_select(target_options_t *options)
 {
   uint32_t dsu_did, id, rev;
+  uint32_t serial_number[4];
   bool locked;
 
   reset_with_extension();
@@ -190,6 +196,14 @@ static void target_select(target_options_t *options)
 
     verbose("Target: %s (Rev %c)\n", devices[i].name, 'A' + rev);
 
+    serial_number[0] = dap_read_word(SERIAL_NUMBER_0);
+    serial_number[1] = dap_read_word(SERIAL_NUMBER_1);
+    serial_number[2] = dap_read_word(SERIAL_NUMBER_2);
+    serial_number[3] = dap_read_word(SERIAL_NUMBER_3);
+
+    verbose("Serial number: %08x %08x %08x %08x\n",
+        serial_number[0], serial_number[1], serial_number[2], serial_number[3]);
+
     target_device = devices[i];
     target_options = *options;
 
-- 
2.14.1

